### PR TITLE
New build platform

### PR DESCRIPTION
**Pull request by razronen**
This PR Contains commonjs & amd release bundles, dev watcher, and gulp deprecation.

### CODE REVIEW

**Comment by lidermanrony** [comment]
Please update readme.md with the latest steps on how to
1. use the package
2. develop for the package (build it ETC)

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,8 @@
 /node_modules/
```
Comment (lidermanrony) [inline .gitignore:1]: looks like this version colors the completion list in blue for some reason when running watch. This is not the case for the older version (and should not be the case for the runtime as well).  Any idea why that is?
image

File: src/languageFeatures.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { LanguageServiceDefaultsImpl } from './monaco.contribution';
 import { KustoWorker } from './kustoWorker';
 
 import * as ls from 'vscode-languageserver-types';
-import * as _ from 'lodash';
+import * as _ from 'lodash-amd';
```
Comment (lidermanrony) [inline src/languageFeatures.ts:5]: why did you need to change this?

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,8 @@
...
 /out/
```
Comment (lidermanrony) [inline .gitignore:2]: general question: were you able to run this in esm mode?

File: package.json [hunk 4]
```diff
@@ -17,30 +17,35 @@
...
     },
     "scripts": {
-        "compile": "gulp compile",
-        "watch": "gulp watch",
-        "prepublish": "gulp release",
-        "test": "jest"
+        "copy_runtime_deps_to_out": "mcopy ./node_modules/@kusto/language-service/kusto.javascript.client.min.js ./out/vs/language/kusto/kusto.javascript.client.min.js && mcopy ./node_modules/@kusto/language-service-next/Kusto.Language.Bridge.min.js ./out/vs/language/kusto/Kusto.Language.Bridge.min.js && mcopy ./node_modules/@kusto/language-service/bridge.min.js ./out/vs/language/kusto/bridge.min.js && node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs && node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs",
+        "compile": "tsc -p ./src/tsconfig.json && tsc -p ./src/tsconfig.esm.json",
+        "watch": "echo dev > test/mode.txt && tsc -p ./src/tsconfig.watch.json && npm run copy_runtime_deps_to_out && concurrently \"live-server ./ \" \"tsc -w -p ./src/tsconfig.watch.json \"",
+        "test_release": "echo release > test/mode.txt && http-server -c-1 -p 8080 ./ -o index.html",
```
Comment (lidermanrony) [inline package.json:23]: looks like http-server is not listed as a dependency. when running the release task i got an error about it.

File: package.json [hunk 5]
```diff
@@ -17,30 +17,35 @@
...
+        "concurrently": "^5.1.0",
+        "copy-dir-cli": "0.0.3",
         "event-stream": "3.3.4",
-        "gulp": "^3.9.1",
-        "gulp-requirejs": "0.1.3",
-        "gulp-tsb": "^2.0.0",
-        "gulp-uglify": "1.5.3",
         "jest": "^22.0.1",
-        "lodash": "^4.17.4",
+        "lodash-amd": "^4.17.15",
```
Comment (lidermanrony) [inline package.json:39]: would that work whne bundling for esm?

File: package.json [hunk 6]
```diff
@@ -17,30 +17,35 @@
...
+        "watch": "echo dev > test/mode.txt && tsc -p ./src/tsconfig.watch.json && npm run copy_runtime_deps_to_out && concurrently \"live-server ./ \" \"tsc -w -p ./src/tsconfig.watch.json \"",
+        "test_release": "echo release > test/mode.txt && http-server -c-1 -p 8080 ./ -o index.html",
+        "copy_types_to_release": "mcopy ./src/monaco.d.ts ./release/min/monaco.d.ts && mcopy ./out/amd/monaco.contribution.d.ts ./release/min/monaco.contribution.d.ts && mcopy ./node_modules/@kusto/language-service/kusto.javascript.client.min.js  ./release/min/kusto.javascript.client.min.js && mcopy ./node_modules/@kusto/language-service-next/Kusto.Language.Bridge.min.js ./release/min/Kusto.Language.Bridge.min.js && mcopy ./node_modules/@kusto/language-service/bridge.min.js ./release/min/bridge.min.js && mcopy ./node_modules/@kusto/language-service/newtonsoft.json.min.js ./release/min/newtonsoft.json.min.js",
+        "release": "mkdirp release  && npm run compile && node ./scripts/release.js && node ./scripts/bundle.js && npm run copy_runtime_deps_to_out && npm run copy_types_to_release && npm run test_release",
```
Comment (lidermanrony) [inline package.json:25]: this looks like it should be in prepublishOnly - since this has to run before npm publish

Comment (lidermanrony) [inline package.json:25]: if this runs on npm publish as per my previous comment - running test_release should not be here (since when publishing we don't need a manual step)

File: package.json [hunk 7]
```diff
@@ -17,30 +17,35 @@
         "url": "https://github.com/Azure/monaco-kusto"
     },
     "scripts": {
-        "compile": "gulp compile",
-        "watch": "gulp watch",
-        "prepublish": "gulp release",
-        "test": "jest"
+        "copy_runtime_deps_to_out": "mcopy ./node_modules/@kusto/language-service/kusto.javascript.client.min.js ./out/vs/language/kusto/kusto.javascript.client.min.js && mcopy ./node_modules/@kusto/language-service-next/Kusto.Language.Bridge.min.js ./out/vs/language/kusto/Kusto.Language.Bridge.min.js && mcopy ./node_modules/@kusto/language-service/bridge.min.js ./out/vs/language/kusto/bridge.min.js && node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs && node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs",
```
[the remark this answers is not in the document]
Comment (razronen) [reply package.json:20]: Where exactly? I couldn't find the duplication

File: package.json [hunk 8]
```diff
@@ -17,30 +17,35 @@
...
+        "compile": "tsc -p ./src/tsconfig.json && tsc -p ./src/tsconfig.esm.json",
```
[the remark this answers is not in the document]
Comment (razronen) [reply package.json:21]: changed to mrmdir

File: package.json [hunk 9]
```diff
@@ -49,7 +54,9 @@
     },
     "dependencies": {
         "@kusto/language-service": "0.0.31",
-        "@kusto/language-service-next": "0.0.34"
+        "@kusto/language-service-next": "0.0.34",
+        "live-server": "^1.2.1",
```
[the remark this answers is not in the document]
Comment (razronen) [reply package.json:58]: sure

File: .gitignore (hunk 3 above)
Comment (razronen) [reply .gitignore:2]: yes

File: package.json (hunk 4 above)
Comment (razronen) [reply package.json:23]: package added

File: package.json (hunk 5 above)
Comment (razronen) [reply package.json:39]: yes worked for me in LogsPortal consuming esm bundles

File: package.json (hunk 6 above)
Comment (razronen) [reply package.json:25]: Right, I don't intend this script to run on npm publish, just as prepublish step

Comment (razronen) [reply package.json:25]: right

File: package.json [hunk 10]
```diff
@@ -17,30 +17,35 @@
         "url": "https://github.com/Azure/monaco-kusto"
```
[the remark this answers is not in the document]
Comment (razronen) [reply package.json:17]: updated

File: .gitignore [hunk 11]
```diff
@@ -1,5 +1,8 @@
...
 /release/
-
+/src/**/*.js.map
+/src/*.js.map
+/test/mode.txt
+package-lock.json
```
[the remark this answers is not in the document]
Comment (razronen) [reply .gitignore:7]: right

File: src/languageFeatures.ts (hunk 2 above)
Comment (razronen) [reply src/languageFeatures.ts:5]: I thought I needed it through the PR cause I didn't find main.js under lodash package, turns out they have lodash.min so I reverted back to regular lodash

File: .gitignore (hunk 1 above)
Comment (razronen) [reply .gitignore:1]: This is really odd.
I see function class in my branch has this color:
image

God knows why. Investigating..

Comment (razronen) [reply .gitignore:1]: It comes from this style element, I have no idea how it gets added to index.html, maybe something to do with not opening the index.html as a static site, and then live-server/http-server pushes this default style element. This happens only when you test the package of course, not consume it.
I added a style element to index.html to override this.
image

File: README.md [hunk 12]
```diff
@@ -97,8 +97,79 @@ Kusto language plugin for the Monaco Editor. It provides the following features
     };
 ```
 
+#### ESM (webpack example)
```
Comment (lidermanrony) [inline README.md:104]: This is a very long list of things to do...
Is there any way to remove steps here by changing the code?

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (lidermanrony) [reply package.json:20]: "copy_runtime_deps_to_out": "mcopy ./node_modules/@kusto/language-service/kusto.javascript.client.min.js ./out/vs/language/kusto/kusto.javascript.client.min.js && mcopy ./node_modules/@kusto/language-service-next/Kusto.Language.Bridge.min.js ./out/vs/language/kusto/Kusto.Language.Bridge.min.js && mcopy ./node_modules/@kusto/language-service/bridge.min.js ./out/vs/language/kusto/bridge.min.js && 
node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs && 
node ./node_modules/copy-dir-cli/bin/copy ./node_modules/monaco-editor-core/dev/vs ./out/vs",

[the remark this answers is not in the document]
Comment (razronen) [reply package.json:20]: Right thanks

File: README.md (hunk 12 above)
Comment (razronen) [reply README.md:104]: Probably some work can be done to make steps 3-4 be smaller,
though this work changes the packing of the package and should be checked carefully against requireJs load. My guideline to this PR was to enable new ESM functionality and to stick with the current packing of AMD modules.

**Comment by noam-shapira** [comment]
merging per @razronen  request
